python/az/aro/azext_aro/custom.py

```diff
@@ -338,3 +303,118 @@ def get_network_resources(cli_ctx, subnets, vnet):
     resources.update(route_tables)
 
     return resources
+
+
+# service_principal_update manages cluster service principal update
+# If called without parameters it should be best-effort
+# If called with parameters it fails if something is not possible
+# Flows:
+# 1. Manual mode where customer provides client_secret and optional client_id
+#      If client_id is provided, we expect expect secret to be provided too
+#      If only secret is provided - we are updating the secret
+# 2. Refresh-cluster-service-principal is provided. client_secret and client_id is not needed
+#      We validate in the validator code so client_id and client_secret is not provided
+#      Check if client_id for existing cluster SP exists for re-usability
+#      If client_id application do not exist - recreate
+#      If SP for client_id do not exist (if we created it in step above it will not) - create
+#  In any case (1,2) we will try to verify and update rbac
+
+def service_principal_update(cli_ctx, oc, client_id=None, client_secret=None, refresh_cluster_service_principal=None):
```

Comment:
This is extremely complicated.  I think we will struggle to support it.

1) Is this because of the AAD weak consistency?

2) Anywhere we join two tasks into one function, we tie ourselves in knots.  One example is create_application setting the password - if we could separate those two without consistency problems, it would simplify our lives.  `generate_secret_by_client_id` is another case in point: IMO it should not call `get_application_by_client_id`.

Bearing in mind the above, if we can find a way around the consistency, could we then pull this into smaller discrete pieces, e.g.:

* if refresh then check or create the app, set the secret, return client_id, client_secret
  else if client_id/client_secret set, check the app, set the secret
* sort out the sp
* validate the resources